flutter-engine/src/texture_registry.rs

```diff
@@ -1,165 +1,185 @@
+        let texture_id = self.texture_id;
+        let frames = self.frames.clone();
+        self.engine.run_on_render_thread(move |engine| {
+            let (width, height) = img.dimensions();
+
+            let glid = unsafe {
+                let mut glid: u32 = 0;
+                gl::GenTextures(1, &mut glid as *mut _);
+                gl::BindTexture(gl::TEXTURE_2D, glid);
+                gl::PixelStorei(gl::UNPACK_ALIGNMENT, 1);
+                gl::TexParameteri(gl::TEXTURE_2D, gl::TEXTURE_MIN_FILTER, gl::LINEAR as _);
+                gl::TexParameteri(gl::TEXTURE_2D, gl::TEXTURE_MAG_FILTER, gl::LINEAR as _);
+                gl::TexImage2D(
+                    gl::TEXTURE_2D,
+                    0,             // mipmap level
+                    gl::RGBA as _, // internal format of the texture
+                    width as _,
+                    height as _,
+                    0,                           // border, must be 0
+                    gl::RGBA,                    // format of the pixel data
+                    gl::UNSIGNED_BYTE,           // data type of the pixel data
+                    (&img).as_ptr() as *const _, // pixel data
+                );
+                gl::BindTexture(gl::TEXTURE_2D, 0);
+
+                glid
+            };
+
+            let engine_weak = engine.downgrade();
+            let frame = TextureFrame::new(gl::TEXTURE_2D, glid, gl::RGBA8, move || {
+                if let Some(engine) = engine_weak.upgrade() {
+                    engine.run_on_render_thread(move |_| unsafe {
+                        gl::DeleteTextures(1, &glid as *const _);
+                    });
+                }
+            });
 
-        let texture = Arc::new(ExternalTexture {
-            texture: flutter_texture,
-            texture_data: Mutex::new(None),
+            post_frame_internal(&engine, texture_id, &frames, frame);
         });
-
-        self.textures
-            .initial
-            .insert(texture_id, Arc::clone(&texture));
-
-        texture
     }
+}
 
-    pub fn get_texture_frame(
-        &mut self,
-        texture_id: TextureID,
-        (width, height): (u32, u32),
-    ) -> Option<ExternalTextureFrame> {
-        if let Some(texture) = self.textures.initial.remove(&texture_id) {
-            // texture is still initial --> create it
-            let frame = unsafe { create_gl_texture(texture_id, (width, height)) };
-            let mut data = texture.texture_data.lock().unwrap();
-            data.replace(TextureData {
-                name: frame.name,
-                width,
-                height,
-            });
-            self.textures
-                .created
-                .insert(texture_id, Arc::downgrade(&texture));
-            return Some(frame);
+fn post_frame_internal(
+    engine: &FlutterEngine,
+    texture_id: TextureId,
+    frames: &Arc<Mutex<HashMap<TextureId, TextureFrame>>>,
+    frame: TextureFrame,
+) {
+    frames.lock().insert(texture_id, frame);
```

Comment:
My bad, I forgot to handle the case where a frame already exists. (two quick calls to post frame for example).

We should use the return value of `insert` and call the destructor callback on the render thread.